embed_grader/serapis/views/tasks.py

```diff
@@ -105,8 +109,46 @@ def delete_assignment_task(request, task_id):
 
     if not user.has_perm('modify_assignment', course):
         return HttpResponse("Not enough privilege")
-    
+
     task.delete()
 
     return HttpResponseRedirect(
             reverse('assignment', kwargs={'assignment_id': assignment.id}))
+
+@login_required(login_url='/login/')
+def zip_input_files(request, task_id):
+    try:
+        task = AssignmentTask.objects.get(id=task_id)
+    except AssignmentTask.DoesNotExist:
+        return HttpResponse("Assignment task cannot be found")
+
+    user = User.objects.get(username=request.user)
+    assignment = task.assignment_fk
+    course = assignment.course_fk
+    if not user.has_perm('view_assignment', course):
+        return HttpResponse("Not enough privilege")
+
+    # retrieve task status
+    # can_see_hidden_cases = (assignment.viewing_scope_by_user(user) == Assignment.VIEWING_SCOPE_FULL)
+    in_memory = BytesIO()
+    input_zip = ZipFile(in_memory, "w")
+    task_files = task.retrieve_assignment_task_files_obj(user)
```

Comment:
Just realize that `retrieve_assignment_task_files_obj()` calls `assignment.viewing_scope_by_user(user)`, but if the user is not enrolled in this course (and hence shouldn't see this assignment), the user can still download the input. Can you help me fix the logic of `viewing_scope_by_user()` in `Assignment` model?

Reply:
I saw that you said you tested a user cannot download input files if she does not enroll in the course. Can you verify which part of the code enforce the permission? It seems the permission check does not happen to be in here...

Reply:
Pls reference line 128-129 in tasks.py, I have put another comment in this PR and explained the concern you mentioned here.

Reply:
Sounds good. Can you do me a favor: Can you still add the permission check in `viewing_scope_by_user()`? If the user doesn't have 'view_assignment' privilege, return `Assignment.VIEWING_SCOPE_NO`

Reply:
Sure, added and committed.